include/zephyr/sys/util.h

```diff
@@ -546,6 +546,55 @@ char *utf8_lcpy(char *dst, const char *src, size_t n);
 		(expr);                                                                            \
 	})
 
+/** @brief Number of bits that make up a type */
+#define NUM_BITS(t) (sizeof(t) * 8)
+
+/**
+ * @brief Compute the log2 of an unsigned integer
+ *
+ * @param x value to compute logarithm of (positive only)
+ *
+ * @returns log2(x)
+ */
+#define LOG2(x)                                                                                    \
+	(_Generic((x), char                                                                        \
+		  : (NUM_BITS(unsigned int) - __builtin_clz(x) - 1), unsigned char                 \
+		  : (NUM_BITS(unsigned int) - __builtin_clz(x) - 1), short                         \
+		  : (NUM_BITS(unsigned int) - __builtin_clz(x) - 1), unsigned short                \
+		  : (NUM_BITS(unsigned int) - __builtin_clz(x) - 1), int                           \
+		  : (NUM_BITS(unsigned int) - __builtin_clz(x) - 1), unsigned int                  \
+		  : (NUM_BITS(unsigned int) - __builtin_clz(x) - 1), long                          \
+		  : (NUM_BITS(unsigned long) - __builtin_clzl(x) - 1), unsigned long               \
+		  : (NUM_BITS(unsigned long) - __builtin_clzl(x) - 1), long long                   \
+		  : (NUM_BITS(unsigned long long) - __builtin_clzll(x) - 1), unsigned long long    \
+		  : (NUM_BITS(unsigned long long) - __builtin_clzll(x) - 1)))
+
+/**
+ * @brief Get the next highest 32-bit power of two for @p x
+ *
+ * This macro rounds-up @p x to the next-highest power of two, as opposed to @ref ROUND_UP which
+ * rounds up a certain value to the next multiple of some alignment.
+ *
+ * @param x the value for which to calculate the next highest power of two
+ *
+ * @retval the next highest power of two for @p x
+ * @retval zero, if the next higher power of two for @p x would saturate 32-bits
+ */
+#define NHPOT(x) (1UL << (LOG2(x) + !IS_POWER_OF_TWO(x)))
```

Comment:
```suggestion
#define NHPOT(x) (1UL << LOG2CEIL(x))
```

Reply:
> I think these can be made more run-time efficient by adding a `LOG2CEIL()` macro:
> LOG2CEIL: `(NUM_BITS(unsigned int) - __builtin_clz(x - 1))`

It looks like there may be an inconsistency for `LOG2CEIL(x)` when `(x == 0)`, but otherwise, seems reasonable. Likely will use Generics again to address the 32-bit / 64-bit problem.

I might move tests to `tests/unit/pot` instead of `tests/unit/nhpot` since there are more now than just the one macro.

Reply:
The other thing I'm curious about, is why `LOG2CEIL()` was added to the `toolchain.h` header. I guess I'll move it and see if it breaks anything 😬

Reply:
> The other thing I'm curious about, is why `LOG2CEIL()` was added to the `toolchain.h` header. I guess I'll move it and see if it breaks anything 😬

I agree. I did search around, and I could not see that it was used anywhere.

Reply:
Here's a test I wrote for it, in case you find use for it:

```
void test_log2ceil(void)
{
	zassert_equal(0, LOG2CEIL(0), NULL);
	zassert_equal(0, LOG2CEIL(1), NULL);
	zassert_equal(1, LOG2CEIL(2), NULL);
	zassert_equal(2, LOG2CEIL(3), NULL);
	zassert_equal(2, LOG2CEIL(4), NULL);
	zassert_equal(3, LOG2CEIL(5), NULL);
	zassert_equal(3, LOG2CEIL(6), NULL);
	zassert_equal(3, LOG2CEIL(7), NULL);
	zassert_equal(3, LOG2CEIL(8), NULL);
	zassert_equal(4, LOG2CEIL(9), NULL);
	zassert_equal(4, LOG2CEIL(10), NULL);
	zassert_equal(4, LOG2CEIL(11), NULL);
	zassert_equal(4, LOG2CEIL(12), NULL);
	zassert_equal(4, LOG2CEIL(13), NULL);
	zassert_equal(4, LOG2CEIL(14), NULL);
	zassert_equal(4, LOG2CEIL(15), NULL);
	zassert_equal(4, LOG2CEIL(16), NULL);
	zassert_equal(5, LOG2CEIL(17), NULL);
	zassert_equal(5, LOG2CEIL(25), NULL);
	zassert_equal(5, LOG2CEIL(32), NULL);
	zassert_equal(6, LOG2CEIL(33), NULL);
	zassert_equal(6, LOG2CEIL(55), NULL);
	zassert_equal(6, LOG2CEIL(64), NULL);

	for (int i = 1; i < 32; i++) {
		zassert_equal(i, LOG2CEIL((1UL << (i - 1)) + 1), NULL);
		zassert_equal(i, LOG2CEIL(1UL << i), NULL);
	}
}
```

Reply:
> It looks like there may be an inconsistency for `LOG2CEIL(x)` when `(x == 0)`

Yes x=0 (and x=1) must be handled separately.

Reply:
I thought the usage in ld files is because LOG2CEIL is a builtin in ld: https://sourceware.org/binutils/docs/ld/Builtin-Functions.html#index-LOG2CEIL_0028exp_0029

Unless you tried removing it and ran into trouble.

Reply:
The LOG2CEIL (now LOG2CEIL_LD) macro was added in d91e4a52de72306b7de6ee1e193120b8dfc75947, as part of https://github.com/zephyrproject-rtos/zephyr/pull/28231 

cc @bbolen

Reply:
Interesting - thanks for pointing out the link. Let me rework it again.